htsget-search/src/htsget/bam_search.rs

```diff
@@ -34,15 +38,49 @@ impl VirtualPositionExt for VirtualPosition {
   /// where that block ends, which is not trivial nor possible for the last block.
   /// The simple solution goes through adding the maximum BGZF block size,
   /// so we don't loose any read (although adding extra unneeded reads to the query results).
-  fn bytes_range_end(&self) -> u64 {
-    self.compressed() + Self::MAX_BLOCK_SIZE
+  fn bytes_range_end(&self, block_ranges: &HashMap<u64, u64>) -> u64 {
+    block_ranges
+      .get(&self.compressed())
+      .cloned()
+      .unwrap_or_else(|| self.compressed() + Self::MAX_BLOCK_SIZE)
   }
 
   fn to_string(&self) -> String {
     format!("{}/{}", self.compressed(), self.uncompressed())
   }
 }
 
+fn load_block_ranges(bai_index: &bai::Index) -> HashMap<u64, u64> {
+  let mut ref_seq_interval: HashMap<u64, u64> =
+    HashMap::with_capacity(bai_index.reference_sequences().len());
+
+  for idx_ref_seq in bai_index.reference_sequences() {
+    if let Some(_) = idx_ref_seq.metadata() {
+      let blocks: HashSet<u64> = idx_ref_seq
+        .bins()
+        .iter()
+        .flat_map(|bin| bin.chunks().iter())
+        .flat_map(|chunk| vec![chunk.start(), chunk.end()])
+        .map(|vpos| vpos.compressed())
+        .collect();
+
+      let mut blocks: Vec<u64> = blocks.into_iter().collect();
+      blocks.sort_unstable();
+
+      let intervals: HashMap<u64, u64> = blocks
+        .iter()
+        .take(blocks.len() - 1)
+        .zip(blocks.iter().skip(1))
+        .map(|(start, end)| (*start, *end))
+        .collect();
+
+      ref_seq_interval.extend(intervals);
+    }
+  }
+
+  ref_seq_interval
+}
```

Comment:
After looking with more care I realized that the code that was suggested in the issue was not accurate/efficient enough.
Please use this other one. I have checked it and there seems to be something else going wrong because there is a test failing. Please could you investigate what is going on?

```suggestion
fn load_block_ranges(bai_index: &bai::Index) -> HashMap<u64, u64> {
  use std::collections::BTreeSet;

  let mut blocks: BTreeSet<u64> = BTreeSet::new();

  for idx_ref_seq in bai_index.reference_sequences() {
    if let Some(_) = idx_ref_seq.metadata() {
      blocks.extend(idx_ref_seq
        .bins()
        .iter()
        .flat_map(|bin| bin.chunks().iter())
        .flat_map(|chunk| vec![chunk.start().compressed(), chunk.end().compressed()])
      );      
    }
  }

  let mut intervals: HashMap<u64, u64> = HashMap::with_capacity(blocks.len());

  intervals.extend(blocks
    .iter()
    .take(blocks.len() - 1)
    .zip(blocks.iter().skip(1))
    .map(|(start, end)| (*start, *end))
  );

  intervals
}
```